refactor: Add `NonNullObjectRef` as part of `ObjectRef`

luukvanderduim opened the pull request:
To encode in the APIs that functions should not be called with a null-reference and to prevent this from happening by accident, we introduce `NonNullObjectRef`.

To prevent repeating ourselves, this refactor changes the `ObjectRef` types to:

```rust
enum ObjectRef<'o> {
    Null,
    NonNull(NonNullObjectRef<'o>),
}

enum NonNullObjectRef<'o> {
    Borrowed { name: UniqueName<'o>, path: ObjectPath<'o> },
    Owned { name: OwnedUniqueName, path: OwnedObjectPath }
}

struct ObjectRefOwned(ObjectRef<'static>)
```

Which offers an `ObjectRef` and `ObjectRefOwned` that can be deserialized from any reference, including null-referenes. Good for receiving references 'bus-side'.

`NonNullObjectRef` is particularly well suited where we want to accept object references.

Addresses: #296

luukvanderduim committed:
refactor: Add `NonNullObjectRef` as part of `ObjectRef`

We'd like to encode in our APIs that some functions should not be
called with a null-reference. This required us to rethink how we
organize our references.

We landed on:

```rust
enum ObjectRef<'o> {
    Null,
    NonNull(NonNullObjectRef<'o>),
}

enum NonNullObjectRef<'o> {
    Borrowed { name: UniqueName<'o>, path: ObjectPath<'o> },
    Owned { name: OwnedUniqueName, path: OwnedObjectPath }
}

struct ObjectRefOwned(ObjectRef<'static>)
```

Which offers us an `ObjectRef` and `ObjectRefOwned` that can be
deserialized from any reference, including null-referenes. Good for
receiving references.

`NonNullObjectRef` is particularly well suited where we want to
accept object references.

luukvanderduim committed:
refactor: Add `NonNullObjectRef` as part of `ObjectRef`

We'd like to encode in our APIs that some functions should not be
called with a null-reference. This required us to rethink how we
organize our references.

We landed on:

```rust
enum ObjectRef<'o> {
    Null,
    NonNull(NonNullObjectRef<'o>),
}

enum NonNullObjectRef<'o> {
    Borrowed { name: UniqueName<'o>, path: ObjectPath<'o> },
    Owned { name: OwnedUniqueName, path: OwnedObjectPath }
}

struct ObjectRefOwned(ObjectRef<'static>)
```

Which offers us an `ObjectRef` and `ObjectRefOwned` that can be
deserialized from any reference, including null-referenes. Good for
receiving references.

`NonNullObjectRef` is particularly well suited where we want to
accept object references.

luukvanderduim committed:
fix(common): fix expected panic message on fail test

luukvanderduim committed:
fix(common): fix expected panic message on fail test

luukvanderduim committed:
doctest(common): a few doctest fixes

and.. remove bad lifetime requirements on return in methods `name()` /
`path()`. We no more need to opt out of clippy.

luukvanderduim committed:
doctest(common): a few doctest fixes

and.. remove bad lifetime requirements on return in methods `name()` /
`path()`. We no more need to opt out of clippy.

luukvanderduim committed:
fix: doctest / lifetime inference in path() / name()

Rustc can only guarantee the static arm if the impl is static too.
But this comes with the disadvantage that a second 'static impl
cannot have same name methods.

I had to reinstate the opt-out for match-on-same-arm pedantic clippy
warning.

fixes doctests

luukvanderduim committed:
fix: doctest / lifetime inference in path() / name()

Rustc can only guarantee the static arm if the impl is static too.
But this comes with the disadvantage that a second 'static impl
cannot have same name methods.

I had to reinstate the opt-out for match-on-same-arm pedantic clippy
warning.

fixes doctests

luukvanderduim committed:
fix: link in doctest

luukvanderduim committed:
fix: link in doctest

luukvanderduim commented:
The fixes to the `text-under-click` example will come in a separate PR.

Fixing it involves a bit of "forensics". I need to recreate the flow as it was when the example ran when we still had null references that could be made into proxies without problems. Performing calls would error, very different from what we have now.

On the bright side the macros I had before are now gone, in favor of mostly this pattern: 

```rust
		// Skip null reference and convert to NonNullObjectRef.
		let Ok(non_null_ref) = NonNullObjectRef::try_from(app.clone()) else { continue };
```

TTWNO commented on atspi-common/src/object_ref.rs:
```diff
@@ -478,27 +770,40 @@ impl PartialEq for ObjectRef<'_> {
+// because PartialEq does not consider it either.
+//
+// This to uphold the contract that if a == b, then a.hash() == b.hash() must hold true.
+impl Hash for NonNullObjectRef<'_> {
```
If you haven't already, please add tests to ensure that this remains the case.

TTWNO commented on atspi/examples/currently-focused-frame.rs:
```diff
@@ -24,15 +24,13 @@ async fn main() -> Result<(), Box<dyn Error>> {
 
-	for app in apps.iter() {
-		let proxy = app.clone().into_accessible_proxy(conn).await?;
+	for app in apps.into_iter().flat_map(|or| NonNullObjectRef::try_from(or).ok()) {
```
On second thought.... this pattern is very common: continue if null, otherwise try to convert to proxy. That should probably just be a method?

TTWNO reviewed (changes requested):
Good work so far. Sorry about my limited free time.

TTWNO commented:
If you are looking at other ways to do the "if null continue, else convert to proxy" thing., you could also try `Option<NonNullObjectRef>::from(ObjectRef).map(...).await?` or something like that :shrug:

luukvanderduim committed:
fix: filtering null in loops

luukvanderduim committed:
fix: filtering null in loops

TTWNO commented:
@luukvanderduim it looks like this PR has a few unrelated fixes in it.

If you can get the git hooks, zbus naming, and the type changes into separate PRs, that would be great.

luukvanderduim commented:
Quick glance in the history and it seems these were separate PR's, so could it be these are here because I rebased to main at some point?
A bit confused why else I would have pulled these in but admittedly have not looked at main yet.

luukvanderduim committed:
refactor(conn|common) Integrating `NonNullObjectRef` and paving the way for zero-copy

The primary objective of this PR is the integration of `NonNullObjectRef` to replace the owned `ObjectRefOwned` object references
in our ongoing effort to better cover where AT-SPI2 might return a NULL reference and where it cannot.
Since object references are ubiquitous across atspi, introducing a lifetime-aware `NonNullObjectRef<'a>`
requires refactoring all over our workspace.

To make this integration possible while laying the groundwork for full zero-copy support in a future PR, several lifetime-decoupling and borrow-checker resolutions were implemented.

- Event<'_> and all 9 wrappers now support retaining the borrow.
- All macros that create events, convert `Message` or convert from or to a wrapper have been adjusted.
  In `MessageConversion` **for now** I chose call `.into_owned()` on the item just to keep the API
  practical and make the transition easier.
  This is the low-hanging fruit. When `Event<'_>::into_owned()` exists (and by extension all wrappers and
  UFET can, then leaving these borrowed will yield a performance win.

- Many tests, examples and benchmarks required adjustments.
- The `event_stream` no longer returns an owned `Event` but returns an owned `Message`.
  Zero-copy an event makes sense if the caller has the owned data under their control.

- Rebased to main
- Fixed tests / doctests
- Solved various doc-issues / clarifications

luukvanderduim committed:
refactor(conn|common) Integrating `NonNullObjectRef` and paving the way for zero-copy

The primary objective of this PR is the integration of `NonNullObjectRef` to replace the owned `ObjectRefOwned` object references
in our ongoing effort to better cover where AT-SPI2 might return a NULL reference and where it cannot.
Since object references are ubiquitous across atspi, introducing a lifetime-aware `NonNullObjectRef<'a>`
requires refactoring all over our workspace.

To make this integration possible while laying the groundwork for full zero-copy support in a future PR, several lifetime-decoupling and borrow-checker resolutions were implemented.

- Event<'_> and all 9 wrappers now support retaining the borrow.
- All macros that create events, convert `Message` or convert from or to a wrapper have been adjusted.
  In `MessageConversion` **for now** I chose call `.into_owned()` on the item just to keep the API
  practical and make the transition easier.
  This is the low-hanging fruit. When `Event<'_>::into_owned()` exists (and by extension all wrappers and
  UFET can, then leaving these borrowed will yield a performance win.

- Many tests, examples and benchmarks required adjustments.
- The `event_stream` no longer returns an owned `Event` but returns an owned `Message`.
  Zero-copy an event makes sense if the caller has the owned data under their control.

- Rebased to main
- Fixed tests / doctests
- Solved various doc-issues / clarifications

luukvanderduim committed:
fix(common): Actually use NULL_OBJECT_NAME_STR

const NULL_OBJECT_NAME_STR is now used.

luukvanderduim committed:
fix(common): Actually use NULL_OBJECT_NAME_STR

const NULL_OBJECT_NAME_STR is now used.

luukvanderduim committed:
docs(common): Link mentioned signals to their respective events

On `NonNullObjectRef` docs link to the events we refer to.

luukvanderduim committed:
docs(common): Link mentioned signals to their respective events

On `NonNullObjectRef` docs link to the events we refer to.

luukvanderduim committed:
fix(common): Elide clippy lint exemptions

By differentiating arms in match arms clippy can now see these
paths are different.

luukvanderduim committed:
fix(common): Elide clippy lint exemptions

By differentiating arms in match arms clippy can now see these
paths are different.

luukvanderduim committed:
tests(common): Equal ObjectRef variants evaluate equal

- Added PartialEq variants that were missing
- Added tests to extensively test:
  - That `ObjectRef` variants, when equal are equal.
  - When `ObjectRef` variants, when equal have equal hashes too.

luukvanderduim committed:
tests(common): Equal ObjectRef variants evaluate equal

- Added PartialEq variants that were missing
- Added tests to extensively test:
  - That `ObjectRef` variants, when equal are equal.
  - When `ObjectRef` variants, when equal have equal hashes too.

luukvanderduim committed:
build(common): Allowing builds with --no-default-features

Feature gate `zbus` was applied too strict, this was found building
wasm32 targets

This problem did not show up for the host build (x86_64) because that
was built using default-features whereas in CI (copied locally) wasm32
builds are performed using --no-default-features.

We gated `NonNullObjectRef` (previously `ObjectRef`) in several files.
Of course that item is `zbus` related, but if not included, event
types won't build.

This disables those gates, making builds with no-default-features succeed.

But how about the altertnative?

Considering the impact of gating the `item` field in event structs would
have had in traits, macros and the API overal - requiring gated special
implementations for non-zbus configurations and special provisioning to
compare, ser/de between platforms, would require A LOT of work for a few
bytes.

luukvanderduim committed:
Clippy(common): Method return acquired an error doc.

luukvanderduim committed:
build: Restore Cargo.lock

luukvanderduim committed:
fix(CI): Fixup direct-direct-minimal-versions

CI pointed out that the transitive requirements on zbus-names and
zvariant requires us to bump versions in the common/Cargo.toml

luukvanderduim committed:
fix(build): Remove local development files

luukvanderduim committed:
build: Restore zbus / zvariant versions

Zbus 5.11.0 had snuck into the workspace, which contains a faulty
DBusError proc-macro impl. 5.12.0 might be too new for some users.

luukvanderduim committed:
fix(conn): Clippy nightly feature combination

We gated import of `Peers` with both wrappers and p2p.
Seems to make little sense(?) let's see.

luukvanderduim committed:
fix(doc): Clippy prefers labal-only if at all possible

luukvanderduim committed:
deps: Restore Cargo.lock from main

toml_parser is a transitive dependency zbus-macros, in zbus-macros
5.15.0 this toml_parser requires Rust 2024. I want to move to 2024
but not in this PR - If I can help it.

TTWNO commented:
6 hours until a job is cancelled... ouch! This does not happen locally I take it?

luukvanderduim committed:
test(conn): doctest must not emit Message from Null path

- Send message with sane path
- Do not fail / continue silently.

On the `event_stream` doctest, we created a Message on the bus
with an origin path "/org/a11y/atspi/null" and we did not fail.
and went on listening.

luukvanderduim commented:
> 6 hours until a job is cancelled... ouch! This does not happen locally I take it?

Well the `event_stream` doctest was "suboptimal" let's say. 
We emitted a message on the bus with a null path origin (bad), we did not time-out ever (also bad) and just continued if the message did not parse into the expected Event, ignoring the error (really bad). ;) (fixed)

luukvanderduim committed:
fix(conn): Remove `wrappers` feature gate on `event_stream`

`event_stream` no longer requires the gate because it no longer
emits `Event` which is a wrapper.

luukvanderduim commented:
# Establish scope for this PR

The PR has overshot its initial goal. 

## How did we end up here?

`NonNullObjectRef<'_>` has a `::Borrowed { .. }` and an `::Owned { .. }` variant, so the enum received a lifetime.
Each UFET needs a `NonNullObjectRef` to tell which object emitted the event. When adding the type to the UFETs, the question arose whether to employ the same trick we had with `ObjectRef<'_>` - wrap it in guaranteed static `ObjectRefOwned(ObjectRef<'static)>` and carry on or realize we have a dot on the horizon involving borrowing from `Message` and prepare for that.
I chose the latter and so this PR does a bit more than its initial goals: 

- supports receiving UFETs with NonNull items on them.
- Event wrappers now support retaining the borrow when wrapping a UFET, 
  meaning the Event<'_> supports being valid for as long as the borrow the UFET has on the `Message`.
- All 9 event wrapping enums now support retaining the the borrow.
- The conversion traits support the borrow on the item, the `NonNullObjectRef`. 
  The work on the body derived UFET fields remains, `(Legacy)CacheItem` being the biggest.
- The `event_stream` now emits `Message`s instead of `Event`s because we cannot lend from the stream.
  This allows users to be flexible on where / when they want to deserialize into an `Event` or 
  `try_into` a UFET.

##  For follow-up PR:

Zero-copy features: 

- All UFETs' fields must have borrowed types with owned / static counterparts
- All UFETs get `.into_owned()` implemented

This will allow users to receive borrowed (zero-copied) event types and turn them into 'static versions
if there is a need to store, send over a channel or an async pattern that requires the data to be static.

- Adjusting the conversion traits and macros implementing these to the borrows for the body derived fields.
  This is now easier than before because the example has been set by the item of each type, so the lifetimes
  are already in place.

## Maybe in this PR: 

Migrate all `ObjectRefOwned` uses in `proxies`' interface argument or return types to `NonNullObjectRef` where applicable and where not. (`Accessible::parent` can't because of, for instance Registry as Accessible::parent), 
I have an a11y tree example somewhere where we can see `Chromium` a11y impl being the odd one at the root node.

I suggest to move that to a follow up PR too as it may require quite a bit of justifying even though we are pretty sure we won't get it all right the first try. 

## Definitely before wrapping up

- A janitorial round at the macros department. Even though macros are said to be hygenic, I do suspect a bit of 
litter around those things.. Names no longer reflect what they do, docs incomplete / outdated, or have TODOs and 
temporary comments.
  Feel free to mark oddities during review!
  
- Find a resolution for the sifting of Null objects from the NonNull objects in the examples. 
The current discussion is a bit of a 'taste debate' TTWNO had in September, I am okay with any solution for now.
Perhaps a cleaner solution is an Iterator adaptor implementation for types whose Deref<Target = &[ObjectRef<'static>])> so we would cover all collections who would have that. Then we could `for child in accessible.get_children().non_null() { ... }

The `NonNull` iterator a idea deservers a bit more time crystalizing perhaps, and it would require a bit of drudgery to go along (testing, docs, doctests).

luukvanderduim committed:
fix(common): Clippy - remove gate on unconditional imports

Previous commit makes `event_stream` unconditional but it depends
on `Stream` and `StreamExt`. Clippy pointed this out, and this
commit fixes it.

T

luukvanderduim committed:
refactor(common): `ObjectRefOwned::into_non_null` conversions

Implements (delegates to From variants):
`ObjectRefOwned::into_non_null`
`ObjectRefOwned::as_non_null`
`ObjectRefNonNull::from_obj_ref`

Implements:
From<ObjectRefOwned> for Option<NonNullObjectRef>
From<ObjectRef<'_> for Option<NonNullObjectRef>
From<&ObjectRefOwned> for Option<&NonNullObjectRef>
From<&ObjectRef<'_> for Option<&NonNullObjectRef>

Plus extensive testing and examples.

THis allows the common pattern in examples:

`for child in children.into_iter().filter_map(ObjectRefOwned::into_non_null)`

luukvanderduim committed:
docs(proxies): Fixup links in `ObjectRefExt` docs

Annotate links to items that are not in scope.

luukvanderduim committed:
CI: fix Codecov GPG signature verification failure

Codecov CLI downloads were failing in the GitHub Actions runner because
the runner lacked Codecov's public PGP key (27034E7FDB850E0BBC2C62FF806BB28AED779869)
in its GPG keyring, leading to a "Can't check signature: No public key" error.

This is a system-level verification issue in the CI runner environment and
is unrelated to any user-specific commit signing keys.

Fixed by importing Codecov's public security keys prior to calling the installer,
or upgrading the workflow step.

luukvanderduim committed:
CI(common): restore zbus message failure-mode tests in macros

Re-add the six failure-mode conversion tests to `zbus_message_test_case`.

(The remaining test-count delta vs. main comes from the new ObjectRefExt
delegation tests and the ObjectRef PartialEq validation tests.)

luukvanderduim committed:
fix(common): fix macro doc drift

Revert `event_wrapper_test_cases` gate to `wrappers` and gate the
`MessageConversion` import on `zbus` so a wrappers-only build still
compiles.

luukvanderduim committed:
fix(common): Fix intra-doc link in macros

luukvanderduim committed:
fix(CI): Undo copied-in 'fix' for coverage

I copied a fix from main to this branch
That fix was later reverted in main.
Rebasing to main reintroduced the 'fix'.

luukvanderduim committed:
docs(proxies): fix doctests HyperLink / EditableText

luukvanderduim committed:
docs(common): Don't refer to private items.

Doctests are free-standing applications that test the public API.
Like integration tests, these cannot see private items, so
CI complains about "items not in scope".

luukvanderduim committed:
revert(All): Restore Cargo.lock to origin/main

Satisfy CI:msrv.

luukvanderduim committed:
fix(common): Don't refer to private item.

luukvanderduim commented:
Comverted to draft.
With the goal to split up the PR.
